vfs: integrate with CJS and ESM module loaders (#63653)
nodejs/node: doc/api/vfs.md

  * `emitExperimentalWarning` {boolean} Whether to emit the experimental
    warning. **Default:** `true`.

### `vfs.mount()`

<!-- YAML
added: REPLACEME
-->

* Returns: {string} The absolute mount point.

Mounts the virtual file system and returns the resulting mount point.
After mounting, files in the VFS can be accessed through the
`node:fs` module and resolved through `require()` and `import`
using paths under the returned mount point.

Mount points always live inside a reserved namespace,
`${os.devNull}/vfs/<layerId>`. Because [`os.devNull`][] is a
character device (POSIX) or a device-namespace path (Windows) that
cannot have child file system entries, no real file-system path can
exist under this namespace: virtual paths never conflate with (or
shadow) real paths, and the layer that owns a path is visible in the

> path itself.

Reviewer: ```suggestion
Mount points always live inside a reserved namespace that cannot have child file system entries,
so virtual paths never conflate with (or shadow) real paths. The virtual path scheme is subject to
change and users should not manually construct them based on assumptions. Instead, obtain
them from what `vfs.mount()` returns or `vfs.mountPoint`.
```

I don't think we should make the path scheme a contract.

overlap.

The VFS supports the [Explicit Resource Management][] proposal. Use
a `using` declaration to unmount automatically when leaving scope:

```cjs
const vfs = require('node:vfs');
const fs = require('node:fs');

let mountPoint;
{
  using myVfs = vfs.create();
  myVfs.writeFileSync('/data.txt', 'Hello');
  mountPoint = myVfs.mount();

  fs.readFileSync(`${mountPoint}/data.txt`, 'utf8'); // 'Hello'
} // VFS is automatically unmounted here

fs.existsSync(`${mountPoint}/data.txt`); // false
```

### `vfs.unmount()`

<!-- YAML
added: REPLACEME
-->

Unmounts the virtual file system. After unmounting, virtual files
are no longer reachable through `node:fs`, `require()`, or `import`.
The same instance may be mounted again by calling `mount()`.

This method is idempotent: calling `unmount()` on a VFS that is not
currently mounted has no effect.

### `vfs.mounted`

<!-- YAML
added: REPLACEME
-->

* {boolean}

`true` while the VFS is mounted; `false` otherwise.

### `vfs.mountPoint`

<!-- YAML
added: REPLACEME
-->

* {string | null}

The current mount point as an absolute string (the value returned by
the last [`vfs.mount()`][] call), or `null` when the VFS is not
mounted.

> ### `vfs.layerId`

Reviewer: Is it really necessary to expose this?

`access`, `rm`, `truncate`, `link`, `mkdtemp`, `chmod`, `chown`, `lchown`,
`utimes`, `lutimes`, `open`, `lchmod`, and `watch`.

## Module loader integration

Once a `VirtualFileSystem` is mounted, paths under the mount point
participate in module resolution and loading. The [CommonJS
resolution algorithm][] used by [`require()`][] and
[`require.resolve()`][] and the [ES modules resolution algorithm][]
used by `import` and [`import.meta.resolve()`][] are unchanged;
instead, every file system operation those algorithms perform
(existence probes, file reads, `package.json` lookups, real-path

> resolution) is dispatched on the path being probed: paths under a

Reviewer: ```suggestion
instead, every file system operation those algorithms perform is dispatched
on the path being probed: paths under a
```

These implementation details should not be enumerated in the docs.

mount point are served by the owning VFS, and all other paths are
served by the real file system. Files served from the VFS therefore
behave as first-class modules: `package.json` is honoured,

> conditional [`"exports"`][] / [`"imports"`][] work, and so on.

Reviewer: ```suggestion
behave as first-class modules.
```

These look like random hand-wavy examples that seem to be meaningful but actually clarify nothing again - if we have to explain with an example it's better to provide a snippet etc. to make it more concrete.

Because mounted paths live in a reserved namespace that cannot exist
on disk, any given path is served either by exactly one VFS or by
the real file system, never both. There is no search order or
fallback between the two: if a path under a mount point does not
exist in the VFS, resolution fails with `ENOENT` without consulting
the disk, and a mounted layer never shadows a real directory.

For resolution purposes the mount point behaves as a file system
root: `package.json` scope lookups (for [`"type"`][],
[`"exports"`][], [`"imports"`][], and the CommonJS directory

> [`"main"`][]) and [loading from `node_modules` folders][] stop at

Reviewer: ```suggestion
root: `package.json` scope lookups and [loading from `node_modules` folders][] stop at
```

Again we should not enumerate them here. If we must, simply linking to the documentation in packages is better.

the mount point instead of continuing into the real file system

> ([the global folders][], such as `NODE_PATH`, are legacy CommonJS

Reviewer: The "instead of continuing into the real file system" part is leading users to spare a thought about a case that is not practical in the first place, because the mount point cannot actually have real files in the file system, no continuation can be performed even if Node.js wants to. I think it's clearer to simply list an example lookup, for example in `/foo/bar/main.cjs`, `require('baz')` should look up

```
- `$MOUNT_POINT/foo/bar/node_modules/baz`
- `$MOUNT_POINT/foo/node_modules/baz`
- `$MOUNT_POINT/node_modules/baz`
- If $NODE_PATH is specified, search folders listed in $NODE_PATH
- `$HOME/.node_modules/baz`
- `$HOME/.node_libraries/baz`
- `$PREFIX/lib/node/baz`
```

Author: I'll need to verify, but I _think_ it's going to recursively to the root of the fs, e.g. down to `/`.

I would not change this specifically for the vfs.

Reviewer: I think in either case, it's better to have this explicit list, otherwise it's not exactly clear from the prose what's the lookup in VFS.

Author: I made the change so that the resolution algorithm stops at vfs mount point, and it does not go back to the root.

behavior and still apply to `require()`). Absolute specifiers may
cross the boundary in either direction: a module on the real file
system can `require()` a mounted path, and a virtual module can
`require()` a real one.

```cjs
const vfs = require('node:vfs');

const myVfs = vfs.create();
myVfs.mkdirSync('/lib');
myVfs.writeFileSync('/lib/greet.js', 'module.exports = () => "hi";');
myVfs.writeFileSync(
  '/lib/package.json', '{"main": "./greet.js"}');
const mountPoint = myVfs.mount();

const greet = require(`${mountPoint}/lib`);
console.log(greet()); // 'hi'

myVfs.unmount();
```

> For ECMAScript modules, convert mounted paths to `file:` URLs before

Reviewer: Non-blocking: I wonder if we should simply supply `vfs.mountPointURL`, then users don't have to convert it...another solution is to provide path methods like `vfs.resolve`/`vfs.resolveAsURL`, then users don't need to go with real path methods to build VFS paths. Also what actually happens under the URL conversion should be much simpler, because there won't be any poking (is it Windows? etc.), it's just simple concatenation.

Author: Yes, I'll add `vfs.mountPointURL`, does not cost us anything and it's more ergonomic.

I'm not sure we should add `resolve`, it's already super-heavy with code and we have that functionality already.

passing them to dynamic `import()`. This keeps VFS imports portable on
Windows, where mounted paths use Windows path syntax.

```mjs
import { pathToFileURL } from 'node:url';
import vfs from 'node:vfs';

const myVfs = vfs.create();
myVfs.writeFileSync('/mod.mjs', 'export const value = 42;');
const mountPoint = myVfs.mount();

const { value } = await import(
  pathToFileURL(`${mountPoint}/mod.mjs`).href,
);
console.log(value); // 42

myVfs.unmount();
```

Module identity follows the path: `__filename` and `module.filename`
are the plain absolute path of the module under the mount point, and
`import.meta.url` is the corresponding `file:` URL, with no synthetic
decorations. Importing the same virtual path repeatedly, including
through `import.meta.resolve()`, yields the same module instance,

> exactly as for real files.

Reviewer: ```suggestion
CommonJS modules loaded from a mounted VFS are identified by their VFS paths
that start with the mount point. This is reflected in, for example, `__filename` and
`__dirname` in the module, or the errors stack traces involving functions from
the VFS modules. ES modules in the VFS are similarly identified by the `file:` URL of
their VFS paths and this is reflected in e.g. `import.meta.url`.

Like modules loaded from the real file system, modules loaded from the VFS are
cached on the first load. When `require()` or `import()` is used to load an absolute
path or URL that falls under the mounted VFS multiple times, the module is only loaded
once and subsequent calls return the same instance.
```

The current description is mixing CommonJS and ESM. It's better to describe different cases more clearly.

Calling [`vfs.unmount()`][] invalidates the modules that were loaded
from the mount point: a subsequent `require()` or `import` of a path
under a re-created mount re-reads the file from the newly mounted
VFS rather than returning a stale module. Modules loaded from other
VFS instances or from the real file system are unaffected.

Mounting and unmounting do not invalidate ESM modules that are
already executing. As with any other module-system teardown,
unmounting a VFS while the import graph below it is still loading is

> the caller's responsibility to avoid.

Reviewer: ```suggestion
Mounting and unmounting do not stop any module execution that is already started,
or invalidate any objects materialized from VFS modules that are already executed.
As with modules in the real file system, the callers are responsible of avoiding
removal or invalidation of modules in the virtual file system while they are being loaded.
```

You could technically unmount CJS while it is executing too (if you make the vfs object reference available e.g. with globals, then call unmount at the top level of a CJS module being loaded from that VFS).

`access`, `rm`, `truncate`, `link`, `mkdtemp`, `chmod`, `chown`, `lchown`,
`utimes`, `lutimes`, `open`, `lchmod`, and `watch`.

## Module loader integration

Once a `VirtualFileSystem` is mounted, paths under the mount prefix
participate in module resolution and loading. Both
`require()` / `require.resolve()` (CommonJS) and `import` /
`import.meta.resolve()` (ECMAScript modules) consult the VFS through
the same toggleable hooks that `node:fs` uses, so files served from
the VFS are first-class modules: `package.json` is honoured,
extensionless files are sniffed for Wasm vs. JavaScript, conditional
`exports` / `imports` work, and so on.

```cjs
const vfs = require('node:vfs');

const myVfs = vfs.create();
myVfs.mkdirSync('/lib');
myVfs.writeFileSync('/lib/greet.js', 'module.exports = () => "hi";');
myVfs.writeFileSync(
  '/lib/package.json', '{"main": "./greet.js"}');
myVfs.mount('/virtual');

const greet = require('/virtual/lib');
console.log(greet()); // 'hi'

myVfs.unmount();
```

### Cache scoping and `import.meta.url`

> Module loaders maintain caches that survive the lifetime of any

Reviewer: I think it's better not to document the cache behavior at length in case this changes and the documented behavior a maintenance burden. Caching should be considered internal details, although module URLs are observable and it's okay to document the changes.

Reviewer: I don't think it's resolved? It's not a very good idea to talk about "the internal stat and
  realpath caches, and the `package.json` caches", those are very much internal details subject to change, and the doc is encouraging people to make assumptions on these internal details..

`access`, `rm`, `truncate`, `link`, `mkdtemp`, `chmod`, `chown`, `lchown`,
`utimes`, `lutimes`, `open`, `lchmod`, and `watch`.

## Module loader integration

Once a `VirtualFileSystem` is mounted, paths under the mount point
participate in module resolution and loading. Both
`require()` / `require.resolve()` (CommonJS) and `import` /
`import.meta.resolve()` (ECMAScript modules) consult the VFS through
the same toggleable hooks that `node:fs` uses, so files served from
the VFS are first-class modules: `package.json` is honoured,
extensionless files are sniffed for Wasm vs. JavaScript, conditional

> `exports` / `imports` work, and so on.

Reviewer: ```suggestion
participate in module resolution and loading. Both
`require()` / `require.resolve()` (CommonJS) and `import` /
`import.meta.resolve()` (ECMAScript modules) consult the VFS through
the same toggleable hooks that `node:fs` uses, so files served from
the VFS are first-class modules: `package.json` is honoured,
extensionless files are sniffed for Wasm vs. JavaScript, conditional
`exports` / `imports` work, and so on.
```

This paragraph is listing example without actually clarifying anything - exactly how do they appear in the module lookup algorithm, and with what precedence?
